pkg/controller/clusterrelocator/clusterrelocator_controller.go

```diff
@@ -0,0 +1,545 @@
+	controllerRelocatorName, relocating := cd.Annotations[constants.RelocatingAnnotation]
+	if !relocating {
+		return nil
+	}
+	logger.WithField("controllerRelocator", controllerRelocatorName).Info("stopping relocation")
+	// TODO: Attempt to clean up resources already relocated
+	delete(cd.Annotations, constants.RelocatingAnnotation)
+	if commitUpdates {
+		if err := r.Update(context.Background(), cd); err != nil {
+			logger.WithError(err).Log(controllerutils.LogLevel(err), "failed to clear relocating annotation")
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *ReconcileClusterRelocator) relocate(cd *hivev1.ClusterDeployment, relocator *hivev1.ClusterRelocator, logger log.FieldLogger) error {
+	logger = logger.WithField("clusterRelocator", relocator.Name)
+
+	if _, relocated := cd.Annotations[constants.RelocatedAnnotation]; !relocated {
+		if cd.Annotations[constants.RelocatingAnnotation] != relocator.Name {
+			if cd.Annotations == nil {
+				cd.Annotations = make(map[string]string, 1)
+			}
+			cd.Annotations[constants.RelocatingAnnotation] = relocator.Name
+			if err := r.Update(context.Background(), cd); err != nil {
+				logger.WithError(err).Log(controllerutils.LogLevel(err), "failed to add relocating annotation")
+				return errors.Wrap(err, "failed to add relocation annotation")
+			}
+		}
+
+		if completed, err := r.move(cd, relocator, logger); err != nil {
+			if conds, changed := controllerutils.SetClusterDeploymentConditionWithChangeCheck(
+				cd.Status.Conditions,
+				hivev1.RelocationFailedCondition,
+				corev1.ConditionTrue,
+				"MoveFailed",
+				err.Error(),
+				controllerutils.UpdateConditionIfReasonOrMessageChange,
+			); changed {
+				cd.Status.Conditions = conds
+				if err := r.Status().Update(context.Background(), cd); err != nil {
+					logger.WithError(err).Log(controllerutils.LogLevel(err), "could not update condition")
+					// return the move error rather than the update error
+				}
+			}
+			return err
+		} else if !completed {
+			return nil
+		}
+
+		delete(cd.Annotations, constants.RelocatingAnnotation)
+		cd.Annotations[constants.RelocatedAnnotation] = relocator.Name
+		if err := r.Update(context.Background(), cd); err != nil {
+			logger.WithError(err).Log(controllerutils.LogLevel(err), "failed to add relocated annotation")
+			return errors.Wrap(err, "failed to add relocated annotation")
+		}
+	}
+
+	if conds, changed := controllerutils.SetClusterDeploymentConditionWithChangeCheck(
+		cd.Status.Conditions,
+		hivev1.RelocationFailedCondition,
+		corev1.ConditionFalse,
+		"MoveSuccessful",
+		"move completed successfully",
+		controllerutils.UpdateConditionNever,
+	); changed {
+		cd.Status.Conditions = conds
+		if err := r.Status().Update(context.Background(), cd); err != nil {
+			logger.WithError(err).Log(controllerutils.LogLevel(err), "could not update condition")
+			return errors.Wrap(err, "could not update condition")
+		}
+	}
+
+	if err := r.Delete(context.Background(), cd); err != nil {
```

Comment:
I can't spot PreserveOnDelete anywhere outside that contrib code, should that not be getting set in this controller?

Reply:
No. The preserve-on-delete functionality is accomplished by the `hive.openshift.io/relocated` annotation. If the ClusterDeployment has been relocated, then the clusterdeployment controller will remove its finalizer without deprovisioning.